packages/contracts-bedrock/contracts/libraries/Types.sol

```diff
@@ -0,0 +1,57 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.9;
+
+/**
+ * @title Types
+ * @notice Contains various types used throughout the Optimism contract system.
+ */
+library Types {
+    /**
+     * @notice OutputProposal represents a commitment to the L2 state. The timestamp is the L1
+     *         timestamp that the output root is posted. This timestamp is used to verify that the
+     *         finalization period has passed since the output root was submitted.
+     */
+    struct OutputProposal {
+        bytes32 outputRoot;
+        uint256 timestamp;
+    }
+
+    /**
+     * @notice Struct representing the elements that are hashed together to generate an output root
+     *         which itself represents a snapshot of the L2 state.
+     */
+    struct OutputRootProof {
+        bytes32 version;
+        bytes32 stateRoot;
+        bytes32 withdrawerStorageRoot;
+        bytes32 latestBlockhash;
+    }
+
+    /**
+     * @notice Struct representing a deposit transaction (L1 => L2 transaction) created by an end
+     *         user (as opposed to a system deposit transaction generated by the system).
+     */
+    struct UserDepositTransaction {
+        address from;
+        address to;
+        uint256 value;
+        uint256 mint;
+        uint64 gasLimit;
+        bool isCreation;
+        bytes data;
+        bytes32 l1BlockHash;
+        uint256 logIndex;
+    }
+
+    /**
+     * @notice Struct representing a withdrawal transaction.
```

Comment:
The nonce here is the nonce in the `L2ToL1MessagePasser` correct? Perhaps good to note so that it can be differentiated from a legacy tx. If solidity had type aliases then we could require that only a specific type of nonce is used here

Reply:
I think this would achieve what you're [looking for](https://docs.soliditylang.org/en/latest/types.html?highlight=custom%20types#user-defined-value-types)?

Reply:
> I think this would achieve what you're [looking for](https://docs.soliditylang.org/en/latest/types.html?highlight=custom%20types#user-defined-value-types)?

Yes it could - I do think introducing type safety for certain interfaces is a good idea